### ISSUE TITLE
libopenshot-audio fails to build under GCC 9.0.1, errors in Juce code

### ISSUE THREAD

**Issue by ferdnyc** [OpenShot/libopenshot-audio#33]
Attempting to build libopenshot-audio for Fedora 30 (which is using gcc 9.0.1), the compilation process is hitting errors in the Juce modules, with an inline function to translate pixel values for the graphics code.

...Code which shouldn't be needed _at all_, but it's there and attempts to manually patch it out have so far proved unsuccessful.

A [full build log](http://koji.rpmfusion.org/kojifiles/work/tasks/511/300511/build.log) is in rpmfusion's koji, but the errors that cause the compilation to fail are:
```
BUILDSTDERR: In file included from /builddir/build/BUILD/libopenshot-audio-0.1.7/JuceLibraryCode/modules/juce_audio_processors/../juce_gui_basics/../juce_graphics/juce_graphics.h:82,
BUILDSTDERR:                  from /builddir/build/BUILD/libopenshot-audio-0.1.7/JuceLibraryCode/modules/juce_audio_processors/../juce_gui_basics/juce_gui_basics.h:28,
BUILDSTDERR:                  from /builddir/build/BUILD/libopenshot-audio-0.1.7/JuceLibraryCode/modules/juce_audio_processors/juce_audio_processors.h:28,
BUILDSTDERR:                  from /builddir/build/BUILD/libopenshot-audio-0.1.7/JuceLibraryCode/modules/juce_audio_processors/juce_audio_processors.cpp:39:
BUILDSTDERR: /builddir/build/BUILD/libopenshot-audio-0.1.7/JuceLibraryCode/modules/juce_audio_processors/../juce_gui_basics/../juce_graphics/colour/juce_PixelFormats.h: In member function 'juce::uint8& juce::PixelARGB::getAlpha()':
BUILDSTDERR: /builddir/build/BUILD/libopenshot-audio-0.1.7/JuceLibraryCode/modules/juce_audio_processors/../juce_gui_basics/../juce_graphics/colour/juce_PixelFormats.h:113:77: error: cannot bind packed field '((juce::PixelARGB*)this)->juce::PixelARGB::<anonymous>.juce::PixelARGB::<unnamed union>::comps[3]' to 'juce::uint8&' {aka 'unsigned char&'}
BUILDSTDERR:   113 |     forcedinline uint8& getAlpha() noexcept           { return comps [indexA]; }
BUILDSTDERR:       |                                                                ~~~~~~~~~~~~~^
BUILDSTDERR: /builddir/build/BUILD/libopenshot-audio-0.1.7/JuceLibraryCode/modules/juce_audio_processors/../juce_gui_basics/../juce_graphics/colour/juce_PixelFormats.h: In member function 'juce::uint8& juce::PixelARGB::getRed()':
BUILDSTDERR: /builddir/build/BUILD/libopenshot-audio-0.1.7/JuceLibraryCode/modules/juce_audio_processors/../juce_gui_basics/../juce_graphics/colour/juce_PixelFormats.h:114:77: error: cannot bind packed field '((juce::PixelARGB*)this)->juce::PixelARGB::<anonymous>.juce::PixelARGB::<unnamed union>::comps[2]' to 'juce::uint8&' {aka 'unsigned char&'}
BUILDSTDERR:   114 |     forcedinline uint8& getRed() noexcept             { return comps [indexR]; }
BUILDSTDERR:       |                                                                ~~~~~~~~~~~~~^
BUILDSTDERR: /builddir/build/BUILD/libopenshot-audio-0.1.7/JuceLibraryCode/modules/juce_audio_processors/../juce_gui_basics/../juce_graphics/colour/juce_PixelFormats.h: In member function 'juce::uint8& juce::PixelARGB::getGreen()':
BUILDSTDERR: /builddir/build/BUILD/libopenshot-audio-0.1.7/JuceLibraryCode/modules/juce_audio_processors/../juce_gui_basics/../juce_graphics/colour/juce_PixelFormats.h:115:77: error: cannot bind packed field '((juce::PixelARGB*)this)->juce::PixelARGB::<anonymous>.juce::PixelARGB::<unnamed union>::comps[1]' to 'juce::uint8&' {aka 'unsigned char&'}
BUILDSTDERR:   115 |     forcedinline uint8& getGreen() noexcept           { return comps [indexG]; }
BUILDSTDERR:       |                                                                ~~~~~~~~~~~~~^
BUILDSTDERR: /builddir/build/BUILD/libopenshot-audio-0.1.7/JuceLibraryCode/modules/juce_audio_processors/../juce_gui_basics/../juce_graphics/colour/juce_PixelFormats.h: In member function 'juce::uint8& juce::PixelARGB::getBlue()':
BUILDSTDERR: /builddir/build/BUILD/libopenshot-audio-0.1.7/JuceLibraryCode/modules/juce_audio_processors/../juce_gui_basics/../juce_graphics/colour/juce_PixelFormats.h:116:77: error: cannot bind packed field '((juce::PixelARGB*)this)->juce::PixelARGB::<anonymous>.juce::PixelARGB::<unnamed union>::comps[0]' to 'juce::uint8&' {aka 'unsigned char&'}
BUILDSTDERR:   116 |     forcedinline uint8& getBlue() noexcept            { return comps [indexB]; }
BUILDSTDERR:       |                                                                ~~~~~~~~~~~~~^
```
(Or, without the insanely wide formatting)
```
juce_PixelFormats.h: In member function 'juce::uint8& juce::PixelARGB::getAlpha()':
juce_PixelFormats.h:113:77: error: cannot bind packed field 
'((juce::PixelARGB*)this)->juce::PixelARGB::<anonymous>.juce::PixelARGB::<unnamed union>::comps[3]'
to 'juce::uint8&' {aka 'unsigned char&'}
   113 |     forcedinline uint8& getAlpha() noexcept           { return comps [indexA]; }
       |                                                                ~~~~~~~~~~~~~^
```

**Comment by ferdnyc** [comment]
Is there any possibility of regenerating the `JuceLibraryCode` directory with the graphical components disabled? I know @RichiCoder1 brought the idea up a while ago in #9 (along with the suggestion since abandoned, that Juce be updated to a newer version) — but now there's a pressing reason to do so, since the existing `juce_graphics` module code doesn't appear to play nice with GCC 9. 

(I suppose there's no guarantee that the modules we _do_ need will all compile cleanly with GCC 9, even if the graphics modules are disabled, but it's at least worth a try since, right now, `libopenshot-audio` is being built with problematic code that it also doesn't actually **need**.)

**Comment by jonoomph** [comment]
@ferdnyc In the past it was not possible. Some of the audio code had dependencies on the graphical components unfortunately. I've tried many times to separate them, but it also ends in defeat, haha. I could regenerate the JuceLibraryCode soon though, and that would probably fix this issue, since this is probably already fixed upstream. But I'm about to branch for our next release, so how about we try this after I branch?

**Comment by ferdnyc** [comment]
@jonoomph That sounds fine, it's not super urgent. Right now we can't build OpenShot for Fedora 30, but the Fedora 29 build should continue to run fine (as far as I know) in the meantime. And, of course, F30 doesn't actually release until the end of April — and that's assuming the schedule doesn't slip — so there's a month and a half left before it really becomes a thing.

**Comment by jonoomph** [comment]
@ferdnyc Still working on this one, and then I'll move on to the other PRs for libopenshot-audio. It's becoming harder to update JUCE these days (they kind of bury the GPLv3 support now).

**Comment by jonoomph** [comment]
I'm documenting this as I go (since I never remember how to do this):
https://github.com/OpenShot/libopenshot-audio/wiki/Update-JUCE-Files

**Comment by ferdnyc** [comment]
@jonoomph 
Cool, I've been playing along at home. Couple of thoughts re: the documentation — which will be great to have, BTW, thanks for putting that together. (I could make the first change directly but since you're updating the page I thought I'd propose it here first, instead.)

#### 1. `AppConfig.h` changes
Currently it says "Update the following line:", but there are two changes and they're not actually adjacent (but could be, I guess). Regardless, another option would be...

> ## 2. Enable GPL Mode (for Projucer)
> In order to manually enable GPLv3-mode for JUCE, we need to edit the following file: `JUCE/extras/Projucer/JuceLibraryCode/AppConfig.h`. Make the following changes:
> ```diff
> --- a/extras/Projucer/JuceLibraryCode/AppConfig.h
> +++ b/extras/Projucer/JuceLibraryCode/AppConfig.h
> @@ -32,7 +32,7 @@
>  // BEGIN SECTION A
>  
>  #ifndef JUCER_ENABLE_GPL_MODE
> - #define JUCER_ENABLE_GPL_MODE 0
> + #define JUCER_ENABLE_GPL_MODE 1
>  #endif
>  
>  // END SECTION A
> @@ -178,7 +178,7 @@
>  // juce_gui_extra flags:
>  
>  #ifndef    JUCE_WEB_BROWSER
> - //#define JUCE_WEB_BROWSER 1
> + #define JUCE_WEB_BROWSER 0
>  #endif
>  
>  
>  #ifndef    JUCE_ENABLE_LIVE_CONSTANT_EDITOR
> ```

The diff takes up a lot more room, yeah, but it's the most unambiguous way to document both changes so maybe it's worth it?

#### 2. `CFLAGS` for freetype2
I noticed that `libopenshot-audio` has always had explicit paths to `freetype2` in its various build files, I always wondered about that. Sort of a tangent, I suppose, but seeing the `CFLAGS` addition on the command line, I got curious all over again.

`freetype2` is supposed to be detected in the Makefile using `pkg-config`, just like all of the other dependencies (which are numerous). If it's installed along with its development files, the include path should be added to `CFLAGS` automatically, since included in the definition for the `JUCE_CPPFLAGS` variable is...
```bash
$(shell pkg-config --cflags freetype2 x11 xext xinerama webkit2gtk-4.0 gtk+-x11-3.0)
```
Can't **hurt** to specify it again, of course, but if that's _necessary_ that would seem like a warning sign that the local system doesn't have a complete installation of the `freetype2` development files. (Projucer built fine for me without any extra `CFLAGS`, anyway. I had to install `webkit2gtk3-devel` first, since `pkg-config` failed to find _that_ library on my system the first time I ran `make`.)

#### 3. `Xinerama`
Skipping ahead a bit, while Projucer depends on `libXinerama`, `libopenshot-audio` doesn't have to — for Fedora I've been patching a `#define JUCE_USE_XINERAMA 0` into `libopenshot-audio/JuceLibraryCode/AppConfig.h` and removing it from `JUCE_PLATFORM_SPECIFIC_LIBRARIES` in the `libopenshot-audio/CMakeFiles.txt`. Those changes are in PR #28, which I know you'd already planned to look at once this is done, but since the Juce configs will be getting regenerated I figured I might mention it now.

**Comment by ferdnyc** [comment]
So, here's an interesting thing I notice.

The modules `juce_graphics`, `juce_gui_basics`, `juce_gui_extra` are required as dependencies, yes... but **only** of the `juce_audio_processors` module, which contains: "Classes for loading and playing VST, AU, LADSPA, or internally-generated audio processors". 

And... it doesn't seem like OpenShot actually _uses_ `AudioProcesssors`. (Not that I can find, anyway.)

The classes that are used, well... 
*  libopenshot's `AudioResampler.cpp`, `AudioReaderSource.cpp`, etc. use things like `ResamplingAudioSource` and `AudioSampleBuffer` (really `AudioBuffer`), those are from `juce_audio_basics`. 
* The `libopenshot-audio/Main.cpp` file uses `AudioDeviceManager` which is part of `juce_audio_devices`. 
* `juce::Thread` and `juce::WaitableEvent` are both used in various headers, they're from `juce_core`.

I'm not even _sure_ `juce_audio_formats` is used, TBH (I can't find where, anyway) — but I'm willing to assume that one must or at least could be useful.

But if `juce_audio_processors` can safely be removed, that eliminates the need to include **any** of the GUI modules, as well as `juce_graphics`. Then, by my count, the required module set is reduced to only:
```xml
        <MODULEPATH id="juce_audio_basics" path="JuceLibraryCode/modules"/>
        <MODULEPATH id="juce_audio_formats" path="JuceLibraryCode/modules"/>
        <MODULEPATH id="juce_core" path="JuceLibraryCode/modules"/>
        <MODULEPATH id="juce_data_structures" path="JuceLibraryCode/modules"/>
        <MODULEPATH id="juce_events" path="JuceLibraryCode/modules"/>
        <MODULEPATH id="juce_audio_devices" path="JuceLibraryCode/modules"/>
```

**Comment by ferdnyc** [comment]
Three things I've discovered, in case they're helpful:

### Module include path
The generated files like `include_juce_audio_basics.cpp` now use e.g. `#include <juce_audio_basics/juce_audio_basics.cpp>` to include the modules, instead of the double-quoted form. Adding the modules dir to the include paths in `CMakeLists.txt` cleared that up for me:
```diff
@@ -139,5 +139,5 @@ ENDIF(WIN32)
 
 # Include header directories
-include_directories("JuceLibraryCode")
+include_directories("JuceLibraryCode" "JuceLibraryCode/modules")
 
 # List of modules to build (Extension based on OS)
```

### Global header file requirement
The module sources also now _**require**_ that an "AppConfig.h" be `#include`d. But since they don't do that _themselves_, my attempts to build the modules kept failing on that check:
```
/usr/lib64/ccache/c++  -DLINUX -Dopenshot_audio_EXPORTS -I/usr/include/freetype2 -I/home/ferd/rpmbuild/REPOS/libopenshot-audio/JuceLibraryCode -I/home/ferd/rpmbuild/REPOS/libopenshot-audio/JuceLibraryCode/modules  -fPIC   -o CMakeFiles/openshot-audio.dir/JuceLibraryCode/modules/juce_audio_basics/juce_audio_basics.cpp.o -c /home/ferd/rpmbuild/REPOS/libopenshot-audio/JuceLibraryCode/modules/juce_audio_basics/juce_audio_basics.cpp
In file included from /home/ferd/rpmbuild/REPOS/libopenshot-audio/JuceLibraryCode/modules/juce_core/juce_core.h:65,
                 from /home/ferd/rpmbuild/REPOS/libopenshot-audio/JuceLibraryCode/modules/juce_audio_basics/juce_audio_basics.h:52,
                 from /home/ferd/rpmbuild/REPOS/libopenshot-audio/JuceLibraryCode/modules/juce_audio_basics/juce_audio_basics.cpp:32:
/home/ferd/rpmbuild/REPOS/libopenshot-audio/JuceLibraryCode/modules/juce_core/system/juce_TargetPlatform.h:56:3: error: #error "No global header file was included!"
  #error "No global header file was included!"
   ^~~~~
make[2]: *** [CMakeFiles/openshot-audio.dir/build.make:63: CMakeFiles/openshot-audio.dir/JuceLibraryCode/modules/juce_audio_basics/juce_audio_basics.cpp.o] Error 1
```

Juce's various header comments and etc. _claim_ that Projucer is supposed to sort that out for you if it's used to generate the files, but it did not seem to — possibly because we're not using their build system to compile. Anyway, I _finally_ fixed that (seemingly) with:
```diff
@@ -177,4 +173,7 @@ TARGET_LINK_LIBRARIES(openshot-audio
 )
 
+target_compile_definitions(openshot-audio PUBLIC
+       JUCE_APP_CONFIG_HEADER="${CMAKE_CURRENT_SOURCE_DIR}/JuceLibraryCode/AppConfig.h")
+
 # PROCESS SUB-DIRECTORIES
 add_subdirectory(src)
```

It _feels_ wrong to have to do that, but it's the only thing I could find that worked.

### Disable `libcurl` dependency
Last but not least, you definitely want to make sure that `AppConfig.h` contains an _explicit_ `#define JUCE_USE_CURL 0`, because `juce_core.h` will explicitly default to 1 under linux:
```h
/** Config: JUCE_USE_CURL
    Enables http/https support via libcurl (Linux only). Enabling this will add an additional
    run-time dynamic dependency to libcurl.

    If you disable this then https/ssl support will not be available on linux.
*/
#ifndef JUCE_USE_CURL
 #if JUCE_LINUX
  #define JUCE_USE_CURL 1
 #else
  #define JUCE_USE_CURL 0
 #endif
#endif
```

**Comment by ferdnyc** [comment]
Aha! Re: the second one, above: It's actually _waaay_ simpler than that.

Juce have reorganized their code to be cleaner, so that now we shouldn't be building any of the files under `JuceLibraryCode/modules/${module_name}` directly. Instead, we should be building `JuceLibraryCode/include_${module_name}.cpp`, which loads the `AppConfig.h` first.

So, basically, the fix is this change to the `CMakeLists.txt`:
```diff
@@ -143,13 +143,7 @@ include_directories("JuceLibraryCode" "JuceLibraryCode/modules")
 
 # List of modules to build (Extension based on OS)
 FILE(GLOB JUCE_SOURCES
-       JuceLibraryCode/modules/juce_core/juce_core.${EXTENSION}
-       JuceLibraryCode/modules/juce_data_structures/juce_data_structures.${EXTENSION}
-       JuceLibraryCode/modules/juce_events/juce_events.${EXTENSION}
-       JuceLibraryCode/modules/juce_audio_basics/juce_audio_basics.${EXTENSION}
-       JuceLibraryCode/modules/juce_audio_devices/juce_audio_devices.${EXTENSION}
-       JuceLibraryCode/modules/juce_audio_formats/juce_audio_formats.${EXTENSION}
-)
+       JuceLibraryCode/include_juce_*.${EXTENSION})
 
 # Disable RPATH
 SET(CMAKE_MACOSX_RPATH 0)
```
That both streamlines the build, _and_ eliminates the messy need to define `JUCE_APP_CONFIG_HEADER`. (`"JuceLibraryCode/modules"` still needs to be added to `INCLUDE_DIRS` though.)

P.S> I'd already removed the lines for the modules I dropped — `juce_audio_processors` and its various graphical dependencies — which is why there are lines missing from the "before" section of the diff, compared to the currently checked-in version.

**Comment by ferdnyc** [comment]
Ohhh-kay. So, I now have "juce5" branches in both of my library forks ([ferdnyc/libopenshot-audio](https://github.com/ferdnyc/libopenshot-audio) and [ferdnyc/libopenshot](https://github.com/ferdnyc/libopenshot-audio)) which I have used to successfully build, package, and install working libraries based on Juce 5.4.3, running on Linux. And, yes, libopenshot-audio now builds under GCC 9 on Fedora 30 as well (but more on that in a second)...

Along the way, I made some pretty major changes to the build environment. The core of those changes were necessitated by the new Juce code structure. (I genuinely believe the new Juce setup is superior, which is why I consider it worth adjusting our build process _to_, rather than trying to swim against the tide in order to preserve legacy cruft.) 

However, along the way I took the opportunity to "spruce up" other aspects of the build in areas where it was lacking. (Like the lack of library path discovery for `libasound` and `libfreetype` in libopenshot-audio, or the convoluted header file structure it outputs, which then became convoluted `#include` structure in `libopenshot`.) The reason I qualified my initial statement with "on Linux" is because, in some instances where my changes necessitate the addition of new packages or libraries to the build configs, I made those additions on the only platform I have access to (Linux), knowing full well that the Windows and Mac builds will break until they've been adjusted as well. 

Which is to say, I explicitly **don't believe** that libopenshot-audio can be compiled under Windows or MacOS using the code on my `juce5` branch. (My changes to libopenshot would be cross-platform, so it should be fine once libopenshot-audio can be built.) The changes required should be _fairly_ minor, but changes will be required. (It should also be fairly simple to selectively undo some of the decisions I made for Linux, as a way of fixing the build on other platforms without having to change their builds. If that approach was preferred.)

Anyway, here's the orbital-satellite view of what I changed in those branches...

#### `libopenshot-audio`
* Generated `JuceLibraryCode` from the Juce 5.4.3 sources, using GPLv3'd Projucer.
    * The modules I mentioned earlier (`audio_processors` and the graphical ones which were only present as its dependencies) are no longer included, which is why compiling under GCC 9 works again: The code that was breaking, in `juce_gui_extra`, is no longer present in the tree.
     * Explicitly set `JUCE_USE_CURL` to 0 in the project settings (as mentioned earlier)
    * Set `JUCE_INCLUDE_ZLIB_CODE` to 0, to use the system zlib instead. 
    This is one of the items that will require Widows and MacOS build changes to compile successfuly on those platforms. However, in addition to being more compliant with the packaging standards for Linux distros, I believe the bundled zlib code may have been the cause of those Windows crashes we were seeing. (Or I could be wrong, and unbundling it from libopenshot-audio could make that situation even worse.) Regardless, every platform has a native zlib shared library trivially accessible, to link in with libopenshot-audio.
* Removed `using namespace std;` from `include/Version.h`, taking to heart OpenShot/libopenshot#204
* Updated the `CMakeLists.txt` build config for the Juce files, so that:
    * The Linux build tree now uses standard CMake `find_package()` modules to detect the configuration for additional dependencies:
        * Freetype (previously hardcoded paths)
        * ALSA (previously no detection, "asound" was  just listed as a library to link in, and assumed to be present)
        * ZLIB (previously not a dependency as it was bundled)
    * Source-file detection for the Juce tree now uses the set of `include_<module>.cpp` files at the root of the `JuceLibraryCode` tree, instead of descending into the `modules/` tree. Juce 5.4.3 is organized in such a way that recursing into `modules/` is no longer supported or necessary.
    * `JuceLibraryCode/modules/` is added to `INCLUDE_DIRECTORIES` when building the source tree. This is necessitated by the use of `#include <juce_$modulename/header.h>` style includes in the `JuceLibraryCode/include_juce_$modulename.cpp` files, rather than the relative-path-based `#include`s used within the `modules/` tree.

        The new code structure requires include paths to be properly set up when building the library itself, just like when building code written against the library, which improves the odds everything will be configured the same way.
    * In concert with the previous change, the Juce header files are now _installed_ at the root of the target include dir. In other words...
        * **Old path:** `/usr/include/libopenshot-audio/JuceLibraryCode/modules/juce_core/juce_core.h`
          **New path:** `/usr/include/libopenshot-audio/juce_core/juce_core.h`
        * **Old path:** `/usr/include/libopenshot-audio/JuceLibraryCode/JuceHeader.h`
          **New path:** `/usr/include/libopenshot-audio/JuceHeader.h`

        This required `#include` changes to a few header files in libopenshot (more on that later), but makes for a much cleaner installation that only requires a single `-I/usr/include/libopenshot-audio` include path to correctly pick up all of the headers.
    * Juce's `AppConfig.h` is also now installed in the `libopenshot-audio` include dir, as it's required when using the Juce headers. (See earlier comments.)

    * Included the `isfinite()` fix (cleaner version) from #19, which is still necessary when building the ruby bindings for libopenshot.

### `libopenshot`
* All header files (7) which contained `#include "JuceLibraryCode/JuceHeader.h"` were changed to simply `#include "JuceHeader.h"`, which works in concert with the new include dir layout for libopenshot-audio.
* Savaged the `cmake/Modules/FindOpenShotAudio.cmake` discovery code to remove all of the recursive header-file discovery and `LIBOPENSHOT_AUDIO_BASE_DIR` noise. 
    Now it simply looks for the file `JuceHeader.h` and considers that location to be the `LIBOPENSHOT_AUDIO_INCLUDE_DIR`, which it is. So much simpler it's not even funny.

I guess I'll submit the branches as PRs to both repos, though as I said the `libopenshot-audio` changes will require (hopefully pretty minor) changes to the Windows and MacOS paths in `CMakeLists.txt` before they can be merged. I don't have the resources to make or test those changes.

**Comment by jonoomph** [comment]
Thanks, I posted a comment on https://github.com/OpenShot/libopenshot-audio/pull/37. I'll work on moving your branches into the OpenShot repo, so our build servers can play with them. Once we get the build servers happy, we can merge this into the `develop` branches.

**Comment by jonoomph** [comment]
This has been merged into `develop`: https://github.com/OpenShot/libopenshot-audio/pull/39
